### PR TITLE
offsetfetch request topics are now nullable

### PR DESCRIPTION
**Pull request by amortezaei**
### Issue
#1115 

### Fix 
made the "Topics" filed of offsetfetch's Request (https://github.com/segmentio/kafka-go/blob/main/protocol/offsetfetch/offsetfetch.go#L11)  nullable in order to make it compliant with Kafka API specs

### Testing

Setup: 

P1 ------> [T1] ------> {CG1} -------> C1
P12------> [T2] ------> {CG1} -------> C2

offset-test-setup

Test 1 driver

```package main

import (
	"context"
	"fmt"

	"github.com/segmentio/kafka-go"
)

func main() {
	fmt.Printf("START\n")
	ctx := context.Background()
	client := kafka.Client{Addr: kafka.TCP("localhost:9092")}
	topicOffsets, err := client.OffsetFetch(ctx, &kafka.OffsetFetchRequest{GroupID: "cg1"})
	if err != nil {
		fmt.Printf("ERROR: %s\n", err.Error())
	}
	fmt.Printf("number of topics %d\n", len(topicOffsets.Topics))
	for topic, offsets := range topicOffsets.Topics {
		fmt.Printf("###########################\n")
		fmt.Printf("offset for topic %s\n", topic)
		for _, offset := range offsets {
			fmt.Printf("\t###########################\n")
			fmt.Printf("\toffset.Partition=%d\n", offset.Partition)
			fmt.Printf("\toffset.CommittedOffset=%d\n", offset.CommittedOffset)
			fmt.Printf("\toffset.Metadata=%s\n", offset.Metadata)
		}
	}

}
```
output:

```
START
number of topics 2
###########################
offset for topic topic-secondary
	###########################
	offset.Partition=5
	offset.CommittedOffset=1349
	offset.Metadata=
	###########################
	offset.Partition=2
	offset.CommittedOffset=912
	offset.Metadata=
	###########################
	offset.Partition=4
	offset.CommittedOffset=228
	offset.Metadata=
	###########################
	offset.Partition=1
	offset.CommittedOffset=685
	offset.Metadata=
	###########################
	offset.Partition=0
	offset.CommittedOffset=687
	offset.Metadata=
	###########################
	offset.Partition=3
	offset.CommittedOffset=686
	offset.Metadata=
###########################
offset for topic topic-primary
	###########################
	offset.Partition=2
	offset.CommittedOffset=525
	offset.Metadata=
	###########################
	offset.Partition=5
	offset.CommittedOffset=1143
	offset.Metadata=
	###########################
	offset.Partition=3
	offset.CommittedOffset=684
	offset.Metadata=
	###########################
	offset.Partition=4
	offset.CommittedOffset=457
	offset.Metadata=
	###########################
	offset.Partition=1
	offset.CommittedOffset=913
	offset.Metadata=
	###########################
	offset.Partition=0
	offset.CommittedOffset=1502
	offset.Metadata=
```

Test 2 driver

```
package main

import (
	"context"
	"fmt"

	"github.com/segmentio/kafka-go"
)

func main() {
	fmt.Printf("START\n")
	ctx := context.Background()
	client := kafka.Client{Addr: kafka.TCP("localhost:9092")}
	topics := map[string][]int{
		"topic-primary": {1, 2, 3},
	}
	topicOffsets, err := client.OffsetFetch(ctx, &kafka.OffsetFetchRequest{GroupID: "cg1", Topics: topics})
	if err != nil {
		fmt.Printf("ERROR: %s\n", err.Error())
	}
	fmt.Printf("number of topics %d\n", len(topicOffsets.Topics))
	for topic, offsets := range topicOffsets.Topics {
		fmt.Printf("###########################\n")
		fmt.Printf("offset for topic %s\n", topic)
		for _, offset := range offsets {
			fmt.Printf("\t###########################\n")
			fmt.Printf("\toffset.Partition=%d\n", offset.Partition)
			fmt.Printf("\toffset.CommittedOffset=%d\n", offset.CommittedOffset)
			fmt.Printf("\toffset.Metadata=%s\n", offset.Metadata)
		}
	}

}
```

output:

```
START
number of topics 1
###########################
offset for topic topic-primary
	###########################
	offset.Partition=1
	offset.CommittedOffset=1360
	offset.Metadata=
	###########################
	offset.Partition=2
	offset.CommittedOffset=525
	offset.Metadata=
	###########################
	offset.Partition=3
	offset.CommittedOffset=912
	offset.Metadata=
```

### CODE REVIEW

File: offsetfetch.go [hunk 1]
```diff
@@ -81,10 +81,20 @@ func (c *Client) OffsetFetch(ctx context.Context, req *OffsetFetchRequest) (*Off
 		})
 	}
 
-	m, err := c.roundTrip(ctx, req.Addr, &offsetfetch.Request{
-		GroupID: req.GroupID,
-		Topics:  topics,
-	})
+	var offsetFetchReq *offsetfetch.Request
+
+	if len(req.Topics) < 1 {
```
Comment (hhahn-tw) [inline offsetfetch.go:86]: Can you add a comment here that topics can be null and kafka will return info on all topics as per spec

File: offsetfetch.go [hunk 2]
```diff
@@ -81,10 +81,25 @@ func (c *Client) OffsetFetch(ctx context.Context, req *OffsetFetchRequest) (*Off
 		})
 	}
 
-	m, err := c.roundTrip(ctx, req.Addr, &offsetfetch.Request{
-		GroupID: req.GroupID,
-		Topics:  topics,
-	})
+	var offsetFetchReq *offsetfetch.Request
```
Comment (petedannemann) [inline offsetfetch.go:84]: What about just changing topics to a nil slice?

e.g. change line 69 to

`var topics []offsetfetch.RequestTopic`

File: offsetfetch.go [hunk 3]
```diff
@@ -66,19 +66,28 @@ type OffsetFetchPartition struct {
...
+	// Kafka version 0.10.2.x and above allow null Topics map for OffsetFetch API
+	// which will return the result for all topics with the desired consumer group:
+	// https://kafka.apache.org/0102/protocol.html#The_Messages_OffsetFetch
+	// For Kafka version below 0.10.2.x this call will result in an error
+	var topics []offsetfetch.RequestTopic
 
-		for i, p := range partitions {
-			indexes[i] = int32(p)
-		}
+	if len(req.Topics) > 0 {
```
Comment (petedannemann) [inline offsetfetch.go:76]: We don't need this:
```go
if len(req.Topics) > 0 {
		topics = make([]offsetfetch.RequestTopic, 0, len(req.Topics))
```
If `req.Topics` is empty then the for loop won't append anything and `topics` will remain nil

Comment (petedannemann) [reply offsetfetch.go:76]: I still see this

**Review by petedannemann — APPROVED** [review]

**Pull request merged**